Fix and improve

Ergus opened the pull request:
- Fixed wrong modeline color issue.
- Added an option to control the verbosity level. Documented in the Readme.

Ergus committed:
Fix mode-line color disable

When canceling with C-g the mode-line color was not reset properly.

Ergus committed:
Add new custom variable to control the message verbosity

Ergus committed:
Small improvement using previous fix.

Ergus committed:
Simplify and improve.

Avoid unneeded funcall and extra call to (syntax-ppss)

Ergus committed:
Change travis test only for emacs versions >= 25.1.

We use funcall-interactively which was introduced in version 25.1.
Also there is the gnutls/emacs issue bug#31313
And newer emacs versions in xenial require update in travis.

Ergus committed:
Add github action to substitute tests

Ergus committed:
Add delete commands to C-k.

Delete is sometimes more useful when we have something in the immediate kill-ring.

Ergus committed:
Separate workflow actions build and test.

Ergus committed:
Add zero as valid numeric argument

Sequences like 03 may be valid.
Untabify.

Ergus committed:
Add some other tests.

Ergus committed:
Simlify composable-goto-char

Rename composable--last-input to composable--char-input

Ergus committed:
Minimal changes

Add some # to function symbols

Ergus committed:
Remove hash table.

It is only faster when there are too many candidates.
In this case an alist is better and simpler to handle.
composable--arguments is a const.

Ergus committed:
Remove unneeded variables and simplify

Ergus committed:
Fix lighter and untabify

Ergus committed:
Fig C-g test.

Ergus committed:
Fix exit with g.

Ergus committed:
Improve and fix test action with cask.

*.github/workflows/test.yml: Use config from cask readme.

Ergus committed:
Add compile command to Makefile.

Ergus committed:
Big changes:

1. Redefine composable-create-composable to use defun with a macro
2. Add copy-region-as-kill-advise to avoid overwriting and create a new
function.
3. Use easy-mmode-defmap for the maps instead of the old syntax.

Ergus committed:
Improve action yml file.

Ergus committed:
Avoid some code duplication.

Ergus committed:
Remove a one use function

Ergus committed:
Remove problematic custom

Ergus committed:
Pacify warning code.

Ergus committed:
Use comment-dwim seems to work

Ergus committed:
Fix delete region binding

Ergus committed:
Fix copy in magit buffers

Ergus committed:
Redefine some bindings

Use [ and ] for begin and end prefix because . must be reserved for
thing-at-point like commands.
Bind . to composable-mark-symbol.

Ergus committed:
Fix overlay.

The composable--overlay was wrong when using commands like select line.

The first selection was fine, but the expansion after that was wrong.

Ergus committed:
Comment always failing test.

The test is actually right. This is mostly an error in cask or the emulator.

Ergus committed:
Delete travis tests

Ergus committed:
Small change

Ergus committed:
Change p for P in interactive

Ergus committed:
Change hopefully unneeded comparison

Ergus committed:
Fix color-supported-p call.

Ergus committed:
Use face remap instead of modifying the modeline-face

Ergus committed:
Remove unneeded variable

paldepind commented on composable.el:
```diff
@@ -238,24 +234,24 @@ For each function named foo a function name composable-foo is created."
 (defun composable--handle-prefix (command)
   "Handle prefix arg where the COMMAND is paired in PAIRS."
-  (let ((pair (gethash command composable--fn-pairs)))
+  (let ((pair (or (alist-get command composable-fn-pair-alist)
```
The old approach of using a hash map seems cleaner to me, and it is more efficient.

Ergus replied:
Hi:

The old approach inserted duplicated elements in the list which is somehow unneeded store '(a b) and '(b a) is redundant and in case we want to distinguish open from close commands is more complex to discriminate. With this, in the future, we can distinguish searching in the forwards `(alist-get command composable-fn-pair-alist)` from searching in the backward `(car (rassq command composable-fn-pair-alist))` if we need that.

But also the customization with the alist is simpler from the user point of view. Because there is not magic function needed; just need to add pairs to the alist. So, this eliminates the need of an extra function to add pairs and exposes this customization variable directly to the user more in the `emacs style` and the user could use either setq, the customize-variable interface, add-to-list and so on.

Finally using a hash table for small arrays like this, is not more efficient than using an alist. Even for big sets it is not so simple and has been discussed many times in the mailing list

https://lists.nongnu.org/archive/html/bug-gnu-emacs/2019-10/msg00779.html 

The hash table overhead is so big that the performance benefit only exists (sometimes) when the set is in the order of the thousands.

paldepind replied:
I see, that makes sense. Thank you for the explanation.

Ergus committed:
Use defsubst for composable-half-cursor.

paldepind commented:
This looks very good to me. There are a few breaking changes though. Maybe it would be good to document those in a changelog?

Ergus commented:
Hi @paldepind :
What do you need to accept this pull request? I have some little time these days to change or modify whatever you need.

BTW: What about adding it to elpa? SOme progress on that?

Ergus committed:
Simplify code that sets the singleton map

Ergus committed:
Simplify composable--activate-repeat.

* composable--repeater: Removed

Ergus committed:
Use cond in composable-object-mode.

This improves readability.

Remove the funcall-interactively use and check in some conditions (this
was a remaining code from a previous version, not useful anymore.)

Ergus committed:
Simplify composable-half-cursor function.

The function just returns the object, don't need to setq cursor-type
because that's already made in the code, so using setq here may be
redundant.

Ergus committed:
;; Style modification.

Use cond instead of if-else in modes. This simplifies readability.

Ergus committed:
Add snapshot action version

Ergus committed:
Try to add Coverals with actions

Ergus committed:
Add new actions

This action should improve coverage, but it seems that the advise we use
is not properly recognized. Any way We keep them.

Ergus committed:
Move back to set-face-background for modeline.

Using face-remap-add-relative breaks some face inheritance. I will ask
about this in the mailing list.
This was exposed by using powerline.

Ergus committed:
Fix keyboard-quit issue.

composable-object-mode should always call (composable--object-exit).
* composable--deactivate-mark-hook-handler : Removed, we don't need it,
it was just a wrapper to go around the interactve-p trick used some time
ago.
* composable-object-mode-disable : Renamed to
composable--object-mode-disable as it is an internal function; use a let
to call composable--object-exit.
* composable--which-key-timer uses a new condition with timerp; this is
saver and prevents errors.

Ergus committed:
Fix previous commit.

Ergus committed:
Fix set-mark-command advice.

Te advice needs to force disable object mode when the mark is
deactivated by calling set-mark command.  This filter is needed because
the mark may be deactivated by other different commands that won't imply
that the composable-object-mode needs to be deactivated.

Ergus committed:
Use a single advice for composable-mark-mode

This simplifies everything because less code handles everything and
solves an small issue experimented with multiple-cursors (actually not supported).

Ergus committed:
Rename finish job to Coveralls.

Ergus committed:
Now is simpler to use an if.

Revert previous change to cond, because a single codition now is smaller
and simpler.

Ergus committed:
Rename advise function to follow the name convention.

* copy-region-as-kill-advise -> composable--copy-region-as-kill-advise

Ergus committed:
"Fix disable composable-object-mode on a second set mark"

Use composable--object-mode-disable because in that case the repeat mode
is enables and the object mode cleanup was not called by the advise when
using (composable-object-mode -1).

Ergus committed:
Try to increase coveralls a bit.

Coveralls has decreased because the code is now much smaller than
before. And some lines can't be tested because are active only when the
display is graphic.

Ergus committed:
Fix composable-mark-mode with multiple cursor.

And also fix set-mark-command-repeat-pop...

Ergus committed:
Not duplicate mark-ring entries on set-mark-command.

Ergus committed:
Simplify set mark condition.

Ergus committed:
Add # to  advice functions.
